Close notification drawer after dismissing last notification

mattmattmatt opened the pull request:
## Proposed change

<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This change adds a listener to any changes of the notifications property once the notification drawer opens. After the last notification is dismissed, the notification drawer closes automatically, and the listener is removed.

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #7223, #4983 
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

mattmattmatt committed:
Close notification drawer after dismissing last

This change adds a listener to any changes of the notifications property once the drawer opens. After the last notification is dismissed, the notification drawer closes automatically, and the listener is removed.

bramkragten commented on src/dialogs/notifications/notification-drawer.js:
```diff
@@ -122,11 +123,31 @@ export class HuiNotificationDrawer extends EventsMixin(
         this.hass.connection,
         (notifications) => {
           this._notificationsBackend = notifications;
+          this.addEventListener(
+            "notifications-changed",
+            this._onNotificationsChanged
+          );
```
We can call this in `_computeNotifications` instead of with an event? And otherwise we should add an `observer` on `notifications` instead of the `notify`

mattmattmatt replied:
The issue with all approaches that don't include an event observer is that the notifications are loaded async. That means when the user tries to _open_ the drawer, for a split second `notifications.length === 0`, and the drawer immediately closes again (actually stays closed).
This is an issue both if we use an `observer` pattern (like `_openChanged` does), and if we move this into `_computeNotifications`. 

While encapsulating everything in `_computeNotifications` is a lot more compact, it runs into the issue mentioned above:
```JS
  _computeNotifications(open, hass, notificationsBackend) {
    if (!open) {
      return [];
    }

    const configuratorEntities = Object.keys(hass.states)
      .filter((entityId) => computeDomain(entityId) === "configurator")
      .map((entityId) => hass.states[entityId]);

    const notifications = notificationsBackend.concat(configuratorEntities);

    // this evaluates to true when the user tries to open the drawer before we have
    // received any notifications from the backend-subscription, thus preventing the drawer from opening
    if (notifications.length === 0) {
        this.open = false;
    }

    return notifications;
  }
```

The only way I have found to avoid this is by attaching the observer once we have loaded all notifications. Another option would be to rewrite the whole component to set up the server subscription in `PolymerElement.ready`, but that would be a larger refactor that I did not feel comfortable proposing for this small feature.

bramkragten replied:
We should rewrite it to a `LitElement`, that will solve your problem and should be done anyway

bramkragten replied:
If you want that challenge that is :-)

mattmattmatt replied:
I have not worked with Polymer or LitElement before, so I'll have to see how deep I want to go into that rabbit hole for adding this mini-feature. 😄 
Until then, is this PR not acceptable?

bramkragten replied:
I would not like to merge it as is, if you want to keep the Polymer you can use:
```js
  static get properties() {
    return {
	  ...,
      notifications: {
        type: Array,
        computed: "_computeNotifications(open, hass, _notificationsBackend)",
        observer: "_notificationsChanged",
      },
      ...,
    };
  }

  _notificationsChanged(newNotifications, oldNotifications) {
    if (this.open && oldNotifications.length && !newNotifications.length) {
      this.open = false;
    }
  }
```

mattmattmatt replied:
I wasn't aware of the `newNotifications, oldNotifications` arguments to the observer function. This is a much easier approach. Thanks for the suggestion!

I adjusted the PR and propose merging like this. If I feel up for it, I'll create a separate PR for a LitElement rewrite, but don't want to commit to it at this time.

mattmattmatt committed:
Use observer instead event for notification change

Using the observer pattern instead subscribing to change events for notification changes simplifies the implementation noticeably.
